_posts/2017-04-04-what-every-developer-should-know-about-ir.md

```diff
@@ -0,0 +1,250 @@
+
+## Vector space vs. boolean model
+
+These models are usually kept clearly separate in research and literature. Most of the real life information retrieval systems are actually hybrid of these using ideas from various sources. For instance predominant Elasticsearch is a hybrid of boolean model and vector space model interpretation. It uses boolean model for matching documents, but utilizes vector space influenced sorting algorithm to sort result set. Postgres has a really good full text search support and it utilizes vector space model pretty cleanly.
+
+Google was originally really simple TF * IDF vector space model system with "page rank" added. Page rank added more weight to pages which are broadly linked from other pages and simirlarly diminished weight of the pages that were seldom linked. For now, we don't know basically anything how Google actually works.
+
+Other models of information retrieval include probabilistic model(s), latent semantic indexing, the binary independence retrieval, best-match retrieval and so on.
+
+## Terms, oh the terms
+
+*Terms* are the atoms of full text information retrieval, and they are called *terms* instead of *words* because usually all kinds of preprocessing is done before indexing and querying. *Terms* come out of *text* with the processes of *tokenisation* and *analysation*.
+
+When the boolean retrieval model was discussed, we already encountered problems with words as they inflect. In boolean model, we addressed to that with OR-facets or using wild cards, but that can be really hard to lazy users who have used to Google. With wild cards one can try to address the problem of word inflection.
+
+There must be something better, right?
+
+We mentioned *query expansion* already and next we take a look at *stop word lists* and *stemming*.
+
+### Stop word list
+
+Usually we just drop words that are not that meaningful in information retrieval context. Stop words in english usually consist of words such as: *a, an, the, none, some, he, she, it, his, her* and so on. When we tokenise raw text, the first thing to do is just drop these and do not add them to index at all. Similarly, when query is processed, the same exact stop word list should be used to filter query as well. It's nice, if user writes query `can I find me a little dog` which effectively queries for `find little dog`.
+
+These words are really common and do not usually give much information about the context of the text. They are just noise, because we have an assumption that term frequencies tell us about the context and content of the texts. It's completely another thing if we try to do something like *question answering*, which the process of finding answers to queries presented in natural language from a collection of texts in natural language. That is really hard in english, but it's almost impossible in finnish!
```

Comment:
that _the_ term frequencies 
*question answering*, which is? the process
hard in _E_nglish ... in _F_innish!